File: src/solutions/2023/10/lua/main.lua
local run = require('src.langs.lua.run')
local misc = require('src.langs.lua.utils.misc')

local options = {
   has_alternate = true,
   has_io = false,
}

local function match_criteria(criteria, c)
   for i = 1, #criteria do
      if string.sub(criteria, i, i) == c then
         return true
      end
   end
   return false
end

local function find_start(grid)
   for y = 1, #grid do
      for x = 1, #grid[y] do
         if grid[y][x] == 'S' then
            return { x, y }
         end
      end
   end
end

local function go_up(grid, x, y)
   local criteria = 'S7F|'
   return y > 1 and match_criteria(criteria, grid[y - 1][x])
end

local function go_down(grid, x, y)
   local criteria = 'SLJ|'
   return y < #grid and match_criteria(criteria, grid[y + 1][x])
end

local function go_left(grid, x, y)
   local criteria = 'SLF-'
   return x > 1 and match_criteria(criteria, grid[y][x - 1])
end

local function go_right(grid, x, y)
   local criteria = 'S7J-'
   return x < #grid[y] and match_criteria(criteria, grid[y][x + 1])
end

local function look_up(grid, x, y)
   local c = grid[y][x]
   local res = {}
   if (c == '|') then
      if (go_up(grid, x, y)) then
         res[#res + 1] = { x, y - 1 }
      end
      if (go_down(grid, x, y)) then
         res[#res + 1] = { x, y + 1 }
      end
   end
   if (c == '-')
   then
      if (go_left(grid, x, y)) then
         res[#res + 1] = { x - 1, y }
      end
      if (go_right(grid, x, y)) then
         res[#res + 1] = { x + 1, y }
      end
   end
   if (c == 'L')
   then
      if (go_up(grid, x, y)) then
         res[#res + 1] = { x, y - 1 }
      end
      if (go_right(grid, x, y)) then
         res[#res + 1] = { x + 1, y }
      end
   end
   if (c == 'J')
   then
      if (go_up(grid, x, y)) then
         res[#res + 1] = { x, y - 1 }
      end
      if (go_left(grid, x, y)) then
         res[#res + 1] = { x - 1, y }
      end
   end
   if (c == '7')
   then
      if (go_down(grid, x, y)) then
         res[#res + 1] = { x, y + 1 }
      end
      if (go_left(grid, x, y)) then
         res[#res + 1] = { x - 1, y }
      end
   end
   if (c == 'F')
   then
      if (go_down(grid, x, y)) then
         res[#res + 1] = { x, y + 1 }
      end
      if (go_right(grid, x, y)) then
         res[#res + 1] = { x + 1, y }
      end
   end
   if (c == 'S')
   then
      if (go_up(grid, x, y)) then
         res[#res + 1] = { x, y - 1 }
      end
      if (go_down(grid, x, y)) then
         res[#res + 1] = { x, y + 1 }
      end
      if (go_left(grid, x, y)) then
         res[#res + 1] = { x - 1, y }
      end
      if (go_right(grid, x, y)) then
         res[#res + 1] = { x + 1, y }
      end
   end
   return res
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local grid = misc.to_grid(input)
   local visited = {}

   for y = 1, #grid do
      visited[y] = {}
      for x = 1, #grid[1] do
         visited[y][x] = false
      end
   end

   local res = 0
   local queue = {}
   queue[#queue + 1] = find_start(grid)
   while #queue > 0 do
      local current = table.remove(queue, #queue)
      local found = look_up(grid, current[1], current[2])
      for _, f in ipairs(found) do
         if not visited[f[2]][f[1]] then
            visited[f[2]][f[1]] = true
            queue[#queue + 1] = f
            res = res + 1
         end
      end
   end

   return tostring(res / 2)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local grid = misc.to_grid(input)
   local visited = {}

   for y = 1, #grid * 3 do
      visited[y] = {}
      for x = 1, #grid[1] * 3 do
         visited[y][x] = false
      end
   end

   local queue = {}
   queue[#queue + 1] = find_start(grid)
   while #queue > 0 do
      local current = table.remove(queue, #queue)
      local found = look_up(grid, current[1], current[2])
      for _, f in ipairs(found) do
         visited[(f[2] - 1) * 3 + 2 + current[2] - f[2]][(f[1] - 1) * 3 + 2 + current[1] - f[1]] = true
         if not visited[(f[2] - 1) * 3 + 2][(f[1] - 1) * 3 + 2] then
            visited[(f[2] - 1) * 3 + 2][(f[1] - 1) * 3 + 2] = true
            queue[#queue + 1] = f
         end
      end
   end

   queue[#queue + 1] = { 1, 1 }
   while #queue > 0 do
      local current = table.remove(queue, #queue)
      if not visited[current[2]][current[1]] then
         visited[current[2]][current[1]] = true
         for ud = -1, 1 do
            for lr = -1, 1 do
               if not (current[2] + ud < 1 or current[2] + ud > #visited or current[1] + lr < 1 or current[1] + lr > #visited[1] or visited[current[2] + ud][current[1] + lr]) then
                  queue[#queue + 1] = { current[1] + lr, current[2] + ud }
               end
            end
         end
      end
   end

   local res = 0
   for y = 1, #grid do
      for x = 1, #grid[1] do
         if not visited[(y - 1) * 3 + 2][(x - 1) * 3 + 2] then
            res = res + 1
         end
      end
   end

   return tostring(res)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()

File: src/solutions/2023/10/python/main.py
import sys
import os
import typing
from collections import deque

sys.path.append(os.getcwd())
from src.langs.python.run import run


def main() -> None:
   options = {'has_alternate': True, 'has_io': False}
   run(sys.argv, part_one, part_two, options)


def find_start(grid: typing.List[str]) -> typing.Tuple[int, int]:
   for y in range(len(grid)):
      for x in range(len(grid[y])):
         if (grid[y][x] == 'S'):
            return (x, y)
   assert "Could not find starting point"


def go_up(grid: typing.List[str], x: int, y: int):
   criteria = 'S7F|'
   return y > 0 and grid[y - 1][x] in criteria


def go_down(grid: typing.List[str], x: int, y: int):
   criteria = 'SLJ|'
   return y < len(grid) - 1 and grid[y + 1][x] in criteria


def go_left(grid: typing.List[str], x: int, y: int):
   criteria = 'SLF-'
   return x > 0 and grid[y][x - 1] in criteria


def go_right(grid: typing.List[str], x: int, y: int):
   criteria = 'S7J-'
   return x < len(grid[y]) - 1 and grid[y][x + 1] in criteria


def look_up(grid: typing.List[str], x: int,
            y: int) -> typing.List[typing.Tuple[int, int]]:
   c = grid[y][x]
   res = []
   if c == '|':
      if go_up(grid, x, y):
         res.append((x, y - 1))
      if go_down(grid, x, y):
         res.append((x, y + 1))
   if c == '-':
      if go_left(grid, x, y):
         res.append((x - 1, y))
      if go_right(grid, x, y):
         res.append((x + 1, y))
   if c == 'L':
      if go_up(grid, x, y):
         res.append((x, y - 1))
      if go_right(grid, x, y):
         res.append((x + 1, y))
   if c == 'J':
      if go_up(grid, x, y):
         res.append((x, y - 1))
      if go_left(grid, x, y):
         res.append((x - 1, y))
   if c == '7':
      if go_down(grid, x, y):
         res.append((x, y + 1))
      if go_left(grid, x, y):
         res.append((x - 1, y))
   if c == 'F':
      if go_down(grid, x, y):
         res.append((x, y + 1))
      if go_right(grid, x, y):
         res.append((x + 1, y))
   if c == 'S':
      if go_up(grid, x, y):
         res.append((x, y - 1))
      if go_down(grid, x, y):
         res.append((x, y + 1))
      if go_left(grid, x, y):
         res.append((x - 1, y))
      if go_right(grid, x, y):
         res.append((x + 1, y))

   return res


def part_one(input: str, is_test: bool) -> str:
   grid = [x for x in input.split('\n')]
   visited = [[False for _ in range(len(grid[0]))] for _ in range(len(grid))]

   i = 0
   queue = deque()
   queue.append(find_start(grid))
   while queue:
      current = queue.popleft()
      found = look_up(grid, current[0], current[1])
      for f in found:
         if (visited[f[1]][f[0]]):
            continue
         visited[f[1]][f[0]] = True
         queue.append(f)
         i += 1

   return str(int(i / 2))


def part_two(input: str, is_test: bool) -> str:
   grid = [x for x in input.split('\n')]
   visited = [[False for _ in range(len(grid[0]) * 3)]
              for _ in range(len(grid) * 3)]

   queue = deque()
   queue.append(find_start(grid))
   while queue:
      current = queue.popleft()
      found = look_up(grid, current[0], current[1])
      for f in found:
         visited[f[1] * 3 + 1 + current[1] - f[1]][f[0] * 3 + 1 + current[0] -
                                                   f[0]] = True
         if (visited[f[1] * 3 + 1][f[0] * 3 + 1]):
            continue
         visited[f[1] * 3 + 1][f[0] * 3 + 1] = True
         queue.append(f)

   queue.append((0, 0))
   while queue:
      current = queue.popleft()
      if (visited[current[1]][current[0]]):
         continue
      visited[current[1]][current[0]] = True
      for ud in [-1, 0, 1]:
         for lr in [-1, 0, 1]:
            if (current[1] + ud < 0 or current[1] + ud >= len(visited)):
               continue
            if (current[0] + lr < 0 or current[0] + lr >= len(visited[0])):
               continue
            if (visited[current[1] + ud][current[0] + lr]):
               continue
            queue.append((current[0] + lr, current[1] + ud))

   res = 0
   for y in range(len(grid)):
      for x in range(len(grid[0])):
         if (not visited[1 + y * 3][1 + x * 3]):
            res += 1

   return str(res)


if __name__ == '__main__':
   main()
